.esh/commands/update-locales.py

```
@@ -1457,15 +1444,53 @@ def generate_ts_type_recursive(data, indent_level):
+		"""Infer a TypeScript array type from a locale list value."""
+		if not arr:
+			return "unknown[]"
+		if all(isinstance(x, str) for x in arr):
+			return "string[]"
+		if all(isinstance(x, (OrderedDict, dict)) for x in arr):
+			return f"Array<{generate_ts_type_recursive(arr[0], indent_level + 1)}>"
+		if all(isinstance(x, list) for x in arr):
+			return f"Array<{generate_ts_array_type(arr[0], indent_level)}>"
+		return "unknown[]"
 
+	def collect_placeholder_keys_recursive(data, prefix=""):
```

Ruff (0.16.0)

[warning] Missing return type annotation for private function `collect_placeholder_keys_recursive`

(ANN202)